### DISCUSSION TITLE
What is the need of QuantumPixel when using MagickConstituteImage

### DISCUSSION THREAD

**Discussion by Haidar0096** [ImageMagick/ImageMagick#5981] [Development]
### ImageMagick version

7

### Operating system, version and so on

windows/android/linux/mac/ios

### Description

`MagickConstituteImage` takes `pixels` array which can be of types `unsigned char, double, float, unsigned int, unsigned long long, unsigned short` and it can also be of `Quantum` type, where `Quantum` is defined according to several factors (Q8, Q16, HDRI, No-HDRI...) and it can be 
```
typedef MagickFloatType Quantum 
typedef unsigned char Quantum 
typedef unsigned short Quantum 
typedef MagickDoubleType Quantum 
typedef unsigned int Quantum
```
my question is that If I understand correctly, then the possible types of `Quantum` are all covered by the types above (`unsigned char, double, float, unsigned int, unsigned long long, unsigned short`), so 
1- why is this type `Quantum` needed? 
2- If I want to make a binding to this method `MagickConstituteImage` then can I drop the option of passing a `Quantum` type of the pixels array? I mean aren't all the possible cases be covered by the other type?

**Comment by snibgo** [comment]
Suppose I was writing a program that loaded pixels from an array, so it would use `MagickConstituteImage()`. I might make the type of array values `long double` to get high precision. But if the IM variety was `Q8 non-hdri` then that precision is overkill: it would massively waste space and time. 

So `MagickConstituteImage()` can use a `StorageType` of `QuantumPixel`. If my program declares the array elements of type `Quantum`, then I know my program will use the precision that IM can provide, without overkill.

[I notice that `StorageType` doesn't include the possibility of `long double`. I don't know why not.]

**Reply by Haidar0096** [reply]
>Suppose I was writing a program that loaded pixels from an array, so it would use MagickConstituteImage(). I might make the type of array values long double to get high precision. But if the IM variety was Q8 non-hdri then that precision is overkill: it would massively waste space and time.

Why would you make the array `long double` if your variant of IM does not support that precision? Don't you know the variant of IM before running the app (i.e. at build time)?

**Comment by snibgo** [comment]
My program can find the variant of IM at build time, of course. And then a series of "`#if ... #elif ... #endif`" could allocate my array accordingly. But it is much simpler to allocate my array with `Quantum` elements. This ensures I haven't forgotten a case, and provides some future proofing, for example if IM one day allows 64-bit integers, or 128-bit integers or 128-bit floating-point.

**Reply by Haidar0096** [reply]
> My program can find the variant of IM at build time, of course. And then a series of "`#if ... #elif ... #endif`" could allocate my array accordingly. But it is much simpler to allocate my array with `Quantum` elements. This ensures I haven't forgotten a case, and provides some future proofing, for example if IM one day allows 64-bit integers, or 128-bit integers or 128-bit floating-point.

I undestand. But still, for the sake of example,  if you are supporting 64-bits and then you "downgrade" you IM variant so it supports only 32 bits, then in this case using `Quantum` type will cause  a transparent loss of information (precision) in the app.

In the end, I want to ask you from your experience, do you have an idea if in practice the `Quantum` type is used alot more than defining the desired type explicitly using other `StorageType`s?

**Comment by snibgo** [comment]
> I undestand. But still, for the sake of example, if you are supporting 64-bits and then you "downgrade" you IM variant so it supports only 32 bits, then in this case using Quantum type will cause a transparent loss of information (precision) in the app.

No. When my program that uses `Quantum` is built, the compiler will read the IM header files, and translate `Quantum` into the appropriate base type. So I can write my program not caring about what data type `Quantum` really is.

> In the end, I want to ask you from your experience, do you have an idea if in practice the Quantum type is used alot more than defining the desired type explicitly using other StorageTypes?

In general, `Quantum` is used far more than explicit data types.

`Quantum` and `QuantumLevel` and `QuantumRange` and others are provided so we can write software that is flexible and resiliant.

**Reply by Piero512** [reply]
What have you seen FFI callers do about this type? The issue stems on the fact that he's trying to use ImageMagick through Dart FFI, and therefore can't know at build-time what type should Quantum be.

**Comment by snibgo** [comment]
I know nothing about Dart. On FFIs in general, see examples at https://imagemagick.org/script/sitemap.php#program-interfaces
